packages/gestalt/src/Typeahead.js

```diff
@@ -0,0 +1,341 @@
+|};
+
+const Typeahead = (props: Props): Node => {
+  const {
+    data,
+    defaultItem = null,
+    id,
+    label = '',
+    noResultText,
+    onBlur,
+    onChange,
+    onFocus,
+    onSelect,
+    placeholder,
+    searchField = 'label',
+    size,
+    forwardedRef,
+  } = props;
+
+  // Store original data
+  const dataRef = useRef(data);
+
+  // Utility function for filtering data by value
+  const filterOriginalData = filterValue =>
+    dataRef.current.filter(item =>
+      item[searchField].toLowerCase().includes(filterValue.toLowerCase())
+    );
+
+  // Track input value
+  const [search, setSearch] = useState<string>(defaultItem?.label || '');
+
+  // Track the selected item - could be used to see if someone is selecting the same thing again
+  const [selected, setSelected] = useState<OptionObject | null>(defaultItem);
+
+  const [hoveredItem, setHoveredItem] = useState<number | null>(0);
+
+  const [options, setOptions] = useState<OptionObject[]>(
+    filterOriginalData(search)
+  );
+
+  // Ref to the input
+  const inputRef = useRef();
+
+  const [containerOpen, setContainerOpen] = useState<boolean>(false);
+
+  const handleFocus = ({ event, value }) => {
+    // Run focus callback
+    if (onFocus) onFocus({ event, value });
+  };
+
+  const handleBlur = ({ event }) => {
+    // Clear input and reset options when no results
+    if (options.length === 0) {
+      setSearch('');
+      setOptions(dataRef.current);
+    }
+
+    setContainerOpen(false);
+
+    // Run blur callback
+    if (onBlur) onBlur({ event });
+  };
+
+  // Handler for when text is typed
+  const handleChange = ({ event, value }) => {
+    if (containerOpen === false) setContainerOpen(true);
+
+    // Filter the available options using original data
+    const updatedOptions = filterOriginalData(value);
+
+    // Update the avalble options
+    setOptions(updatedOptions);
+
+    // Update the search value
+    setSearch(value);
+
+    // Run onChange callback
+    if (onChange) onChange({ event, value });
+  };
+
+  const handleClear = () => {
+    setSelected(null);
+    setSearch('');
+    setContainerOpen(false);
+    if (inputRef.current) inputRef.current.focus();
+  };
+
+  // Handler for when an item is clicked
+  const handleSelect = ({ event, item }) => {
+    setSelected(item);
+
+    setSearch(item[searchField]);
+    setContainerOpen(false);
+    if (inputRef.current) inputRef.current.focus();
+    if (onSelect) onSelect({ event, item });
+  };
+
+  let selectedElement;
+  const setOptionRef = ref => {
+    selectedElement = ref;
+  };
+
+  const containerRef = useRef();
+
+  const handleScrolling = (direction: number) => {
+    const container = containerRef.current;
+
+    // Based on keyboard navigation we get the next or previuos option
+    // When we reach the start or end of the list, move to the start or end of the list based on the direction
+    const nextOption =
+      direction > 0
+        ? // $FlowFixMe[incompatible-cast]
```

Comment:
@simkessy could you explain why we need casting here + can we get rid of the `// $FlowFixMe`?

Reply:
the nextSibling callback returns a type of Node however nextOption is should be htmlElement